### ISSUE TITLE
Config key not working when used together with setup key

### ISSUE THREAD

**Issue by datwaft** [wbthomason/packer.nvim#92]
Hello, I am trying to configure my _colorscheme_ using the `config` and `setup` keys but the `config` function doesn't execute, only the `setup` function; if I only have `config` in plugin it executes, though.

I have executed `:PackerCompile` many times, also.

I am testing this behaviour with `:echom` and `:messages`.

Here is a screenshot:

![image](https://user-images.githubusercontent.com/37723586/99738784-e759a080-2a90-11eb-925b-3b1650eb19fe.png)

## Environment

Ubuntu-20.04 WSL 2 Windows 10

```viml
:version
NVIM v0.5.0-843-gce70edb80
Build type: Release
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/cc -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -DNDEBUG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-pr
ototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIB
I_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/datwaft/downloads/neovim/build/config -I/home/datwaft/downloads/neovim/src -I/home/datwaft/downloads/neovim/.deps/usr/include -I/usr/include -I/ho
me/datwaft/downloads/neovim/build/src/nvim/auto -I/home/datwaft/downloads/neovim/build/include
Compiled by datwaft@DESKTOP-FRA5VGQ

Features: +acl +iconv +tui
See ":help feature-compile"

  system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/share/nvim"

Run :checkhealth for more info
```

## What do I expect to happen

```viml
:messages
loading setup...
loading config...
```

## What actually happens

```viml
:messages
loading setup...
```

## Relevant configuration

`init.vim`
```viml
lua require('plugins')
```

`lua/plugins.lua`
```lua
vim.cmd [[packadd packer.nvim]]

return require('packer').startup(function()
   use {'wbthomason/packer.nvim', opt = true}
   use {
      'bluz71/vim-nightfly-guicolors',
      config = require('plugins.colorscheme').config,
      setup = require('plugins.colorscheme').setup
   }
end)
```

`lua/plugins/colorscheme.lua`
```lua
return {
   config = function()
      vim.cmd [[ echom "loading config..." ]]
      vim.g.nightflyCursorColor = 1
      vim.cmd [[ colorscheme nightfly ]]
   end,
   setup = function()
      vim.cmd [[ echom "loading setup..." ]]
      vim.o.termguicolors = true
   end
}
```

## Notes

Already tried without my `configuration.lua` file and without any other plugin other than `packer.nvim` and my _colorscheme_.

**Comment by datwaft** [comment]
Additionally, currently I decide to just use a config key with everything there.

![image](https://user-images.githubusercontent.com/37723586/99739543-9d71ba00-2a92-11eb-8e9d-e983beb4b7d7.png)

**Comment by wbthomason** [comment]
Could you please post your `packer_compiled.vim` file? I'd like to inspect it; I have a suspicion about a possible bug.

**Comment by datwaft** [comment]
Here is my current `packer-compiled.vim`:  [current-file.zip](https://github.com/wbthomason/packer.nvim/files/5576523/current-file.zip)

<details>
<summary>Current file</summary>

```viml
" Automatically generated packer.nvim plugin loader code

if !has('nvim')
  finish
endif

lua << END
local plugins = {
  ["packer.nvim"] = {
    loaded = false,
    only_sequence = false,
    only_setup = false,
    path = "/home/datwaft/.local/share/nvim/site/pack/packer/opt/packer.nvim"
  }
}

local function handle_bufread(names)
  for _, name in ipairs(names) do
    local path = plugins[name].path
    for _, dir in ipairs({ 'ftdetect', 'ftplugin', 'after/ftdetect', 'after/ftplugin' }) do
      if #vim.fn.finddir(dir, path) > 0 then
        vim.cmd('doautocmd BufRead')
        return
      end
    end
  end
end

_packer_load = nil

local function handle_after(name, before)
  local plugin = plugins[name]
  plugin.load_after[before] = nil
  if next(plugin.load_after) == nil then
    _packer_load({name}, {})
  end
end

_packer_load = function(names, cause)
  local some_unloaded = false
  for _, name in ipairs(names) do
    if not plugins[name].loaded then
      some_unloaded = true
      break
    end
  end

  if not some_unloaded then return end

  local fmt = string.format
  local del_cmds = {}
  local del_maps = {}
  for _, name in ipairs(names) do
    if plugins[name].commands then
      for _, cmd in ipairs(plugins[name].commands) do
        del_cmds[cmd] = true
      end
    end

    if plugins[name].keys then
      for _, key in ipairs(plugins[name].keys) do
        del_maps[key] = true
      end
    end
  end

  for cmd, _ in pairs(del_cmds) do
    vim.cmd('silent! delcommand ' .. cmd)
  end

  for key, _ in pairs(del_maps) do
    vim.cmd(fmt('silent! %sunmap %s', key[1], key[2]))
  end

  for _, name in ipairs(names) do
    if not plugins[name].loaded then
      vim.cmd('packadd ' .. name)
      if plugins[name].config then
        for _i, config_line in ipairs(plugins[name].config) do
          loadstring(config_line)()
        end
      end

      if plugins[name].after then
        for _, after_name in ipairs(plugins[name].after) do
          handle_after(after_name, name)
          vim.cmd('redraw')
        end
      end

      plugins[name].loaded = true
    end
  end

  handle_bufread(names)

  if cause.cmd then
    local lines = cause.l1 == cause.l2 and '' or (cause.l1 .. ',' .. cause.l2)
    vim.cmd(fmt('%s%s%s %s', lines, cause.cmd, cause.bang, cause.args))
  elseif cause.keys then
    local keys = cause.keys
    local extra = ''
    while true do
      local c = vim.fn.getchar(0)
      if c == 0 then break end
      extra = extra .. vim.fn.nr2char(c)
    end

    if cause.prefix then
      local prefix = vim.v.count and vim.v.count or ''
      prefix = prefix .. '"' .. vim.v.register .. cause.prefix
      if vim.fn.mode('full') == 'no' then
        if vim.v.operator == 'c' then
          prefix = '' .. prefix
        end

        prefix = prefix .. vim.v.operator
      end

      vim.fn.feedkeys(prefix, 'n')
    end

    -- NOTE: I'm not sure if the below substitution is correct; it might correspond to the literal
    -- characters \<Plug> rather than the special <Plug> key.
    vim.fn.feedkeys(string.gsub(string.gsub(cause.keys, '^<Plug>', '\\<Plug>') .. extra, '<[cC][rR]>', '\r'))
  elseif cause.event then
    vim.cmd(fmt('doautocmd <nomodeline> %s', cause.event))
  elseif cause.ft then
    vim.cmd(fmt('doautocmd <nomodeline> %s FileType %s', 'filetypeplugin', cause.ft))
    vim.cmd(fmt('doautocmd <nomodeline> %s FileType %s', 'filetypeindent', cause.ft))
  end
end

-- Runtimepath customization

-- Pre-load configuration
-- Post-load configuration
-- Config for: vim-nightfly-guicolors
loadstring("\27LJ\2\2\1\0\0\2\0\a\0\r6\0\0\0009\0\1\0+\1\2\0=\1\2\0006\0\0\0009\0\3\0)\1\1\0=\1\4\0006\0\0\0009\0\5\0'\1\6\0B\0\2\1K\0\1\0\27 colorscheme nightfly \bcmd\24nightflyCursorColor\6g\18termguicolors\6o\bvim\0")()
-- Config for: coc.nvim
loadstring("\27LJ\2\2¦\1\0\0\3\0\b\0\t6\0\0\0009\0\1\0005\1\3\0=\1\2\0006\0\4\0009\0\5\0'\1\6\0'\2\a\0D\0\3\0\29:CocCommand explorer<CR>\n<C-n>\tnmap\tvimp\1\5\0\0\20coc-marketplace\17coc-explorer\17coc-snippets\fcoc-lua\26coc_global_extensions\6g\bvim\0")()
-- Conditional loads
-- Load plugins in order defined by `after`
END

function! s:load(names, cause) abort
call luaeval('_packer_load(_A[1], _A[2])', [a:names, a:cause])
endfunction


" Command lazy-loads

" Keymap lazy-loads

augroup packer_load_aucmds
  au!
  " Filetype lazy-loads
  " Event lazy-loads
augroup END
```

</details>

Here is a `packer-compiled.vim` trying to recreate the bug:  [bug-recreated.zip](https://github.com/wbthomason/packer.nvim/files/5576565/bug-recreated.zip)


<details>
<summary>File trying to recreate the bug</summary>

```viml
" Automatically generated packer.nvim plugin loader code

if !has('nvim')
  finish
endif

lua << END
local plugins = {
  ["packer.nvim"] = {
    loaded = false,
    only_sequence = false,
    only_setup = false,
    path = "/home/datwaft/.local/share/nvim/site/pack/packer/opt/packer.nvim"
  },
  ["vim-nightfly-guicolors"] = {
    config = { "\27LJ\2\0024\0\0\2\0\3\0\0056\0\0\0009\0\1\0'\1\2\0B\0\2\1K\0\1\0\21 echom \"config\" \bcmd\bvim\0" },
    loaded = false,
    only_sequence = true,
    only_setup = true,
    path = "/home/datwaft/.local/share/nvim/site/pack/packer/opt/vim-nightfly-guicolors"
  }
}

local function handle_bufread(names)
  for _, name in ipairs(names) do
    local path = plugins[name].path
    for _, dir in ipairs({ 'ftdetect', 'ftplugin', 'after/ftdetect', 'after/ftplugin' }) do
      if #vim.fn.finddir(dir, path) > 0 then
        vim.cmd('doautocmd BufRead')
        return
      end
    end
  end
end

_packer_load = nil

local function handle_after(name, before)
  local plugin = plugins[name]
  plugin.load_after[before] = nil
  if next(plugin.load_after) == nil then
    _packer_load({name}, {})
  end
end

_packer_load = function(names, cause)
  local some_unloaded = false
  for _, name in ipairs(names) do
    if not plugins[name].loaded then
      some_unloaded = true
      break
    end
  end

  if not some_unloaded then return end

  local fmt = string.format
  local del_cmds = {}
  local del_maps = {}
  for _, name in ipairs(names) do
    if plugins[name].commands then
      for _, cmd in ipairs(plugins[name].commands) do
        del_cmds[cmd] = true
      end
    end

    if plugins[name].keys then
      for _, key in ipairs(plugins[name].keys) do
        del_maps[key] = true
      end
    end
  end

  for cmd, _ in pairs(del_cmds) do
    vim.cmd('silent! delcommand ' .. cmd)
  end

  for key, _ in pairs(del_maps) do
    vim.cmd(fmt('silent! %sunmap %s', key[1], key[2]))
  end

  for _, name in ipairs(names) do
    if not plugins[name].loaded then
      vim.cmd('packadd ' .. name)
      if plugins[name].config then
        for _i, config_line in ipairs(plugins[name].config) do
          loadstring(config_line)()
        end
      end

      if plugins[name].after then
        for _, after_name in ipairs(plugins[name].after) do
          handle_after(after_name, name)
          vim.cmd('redraw')
        end
      end

      plugins[name].loaded = true
    end
  end

  handle_bufread(names)

  if cause.cmd then
    local lines = cause.l1 == cause.l2 and '' or (cause.l1 .. ',' .. cause.l2)
    vim.cmd(fmt('%s%s%s %s', lines, cause.cmd, cause.bang, cause.args))
  elseif cause.keys then
    local keys = cause.keys
    local extra = ''
    while true do
      local c = vim.fn.getchar(0)
      if c == 0 then break end
      extra = extra .. vim.fn.nr2char(c)
    end

    if cause.prefix then
      local prefix = vim.v.count and vim.v.count or ''
      prefix = prefix .. '"' .. vim.v.register .. cause.prefix
      if vim.fn.mode('full') == 'no' then
        if vim.v.operator == 'c' then
          prefix = '' .. prefix
        end

        prefix = prefix .. vim.v.operator
      end

      vim.fn.feedkeys(prefix, 'n')
    end

    -- NOTE: I'm not sure if the below substitution is correct; it might correspond to the literal
    -- characters \<Plug> rather than the special <Plug> key.
    vim.fn.feedkeys(string.gsub(string.gsub(cause.keys, '^<Plug>', '\\<Plug>') .. extra, '<[cC][rR]>', '\r'))
  elseif cause.event then
    vim.cmd(fmt('doautocmd <nomodeline> %s', cause.event))
  elseif cause.ft then
    vim.cmd(fmt('doautocmd <nomodeline> %s FileType %s', 'filetypeplugin', cause.ft))
    vim.cmd(fmt('doautocmd <nomodeline> %s FileType %s', 'filetypeindent', cause.ft))
  end
end

-- Runtimepath customization

-- Pre-load configuration
-- Setup for: vim-nightfly-guicolors
loadstring("\27LJ\2\0023\0\0\2\0\3\0\0056\0\0\0009\0\1\0'\1\2\0B\0\2\1K\0\1\0\20 echom \"setup\" \bcmd\bvim\0")()
vim.cmd("packadd vim-nightfly-guicolors")
-- Post-load configuration
-- Config for: coc.nvim
loadstring("\27LJ\2\2¦\1\0\0\3\0\b\0\t6\0\0\0009\0\1\0005\1\3\0=\1\2\0006\0\4\0009\0\5\0'\1\6\0'\2\a\0D\0\3\0\29:CocCommand explorer<CR>\n<C-n>\tnmap\tvimp\1\5\0\0\20coc-marketplace\17coc-explorer\17coc-snippets\fcoc-lua\26coc_global_extensions\6g\bvim\0")()
-- Conditional loads
-- Load plugins in order defined by `after`
END

function! s:load(names, cause) abort
call luaeval('_packer_load(_A[1], _A[2])', [a:names, a:cause])
endfunction


" Command lazy-loads

" Keymap lazy-loads

augroup packer_load_aucmds
  au!
  " Filetype lazy-loads
  " Event lazy-loads
augroup END
```

</details>

**Comment by wbthomason** [comment]
Thanks for this report! It revealed a logic error in the lazy-load compiler. This should be fixed by 3a4fe6b; please let me know (and reopen the issue) if you still have the original problem or notice any regressions.
